### ISSUE TITLE
Script biorxiv preprint dataset creation

### ISSUE THREAD

**Issue by dhimmel** [dhimmel/biorxiv-licenses#1]
@OmnesRes initially provided us with `biorxiv_data.txt` [via Twitter](https://twitter.com/OmnesResNetwork/status/802726795128795136) (in 54db0731a6dae6ca3ed60474d8bbbf67a5c8e058). In a series of commits, I modified this file to create [`biorxiv-prepubmed.tsv`](https://github.com/dhimmel/biorxiv-licenses/blob/388749ca1bdc92b17d3ac43121d35c4e0a787fd6/data/biorxiv-prepubmed.tsv) which uses a simpler format. It would be great to script the creation of `biorxiv-prepubmed.tsv` so we could run in the future and make this study reproducible.

@OmnesRes is the generation of this dataset possible via public APIs? Do you have code that you used to create `biorxiv_data.txt`? I'm happy to help you create a pull request to add this functionality.

**Comment by OmnesRes** [comment]
I don't believe bioRxiv has an API.  I'm using Python requests and BeautifulSoup to scrape their site.  This is the initial script PrePubMed used to index articles in bioRxiv [link](https://github.com/OmnesRes/prepub/blob/master/biorxiv/biorxiv.py).  Every day my web server indexes new articles and they get added to this [file](https://github.com/OmnesRes/prepub/blob/master/biorxiv/biorxiv.txt).  This file contains all the articles stored as Python lists in this order: [title, author, date, abstract, link, subjects, author affiliations].  When you requested the licenses I used the links from this file to visit the .article-info of each page which is where the license information is located.  I then merged this information with the date, link (doi), and subject information I already had.  PrePubMed seems to be missing a lot of subjects for bioRxiv.  I think when they post an article the subject might not be present so PrePubMed misses it on its daily indexing, so I also went ahead and grabbed missing subjects for you.

This Python script: https://s3.amazonaws.com/omnesres/biorxiv_licenses.py will take the articles in biorxiv.txt, see which articles are missing in your file biorxiv-licenses.tsv, then grab their license information and check for missing subject information.  It will then write a file just like the one I shared on Twitter.  I guess I could have formatted the file like yours, but it seems you have code for that already.

It seems the question is how often do you want to run biorxiv_licenses.py?  I commit to the https://github.com/OmnesRes/prepub repository every couple days.  If you fork that repository you'll be notified when there are potentially new bioRxiv articles that need their licensing information added to your file.

**Comment by OmnesRes** [comment]
By the way, were you planning on doing some analyses on the data?  I would be interested in how license preferences have changed over time or how license preferences differ among the subject areas.  If you include author information you could also check if authors are consistent with the licenses they select.

**Comment by dhimmel** [comment]
>  I would be interested in how license preferences have changed over time or how license preferences differ among the subject areas.

Exactly what I was thinking. Will get to this after I get altair/vega working in mybinder.org notebooks (#2). Was thinking either a [streamgraph](https://altair-viz.github.io/gallery/stacked_area_stream.html), [Normalized Stacked Area Chart](https://altair-viz.github.io/gallery/stacked_area_normalize.html#gallery-stacked-area-normalize), or Stacked Area Chart.

> If you include author information you could also check if authors are consistent with the licenses they select.

Great idea!

For new ideas, how about opening new GitHub issues. Just want to keep this clean.

**Comment by dhimmel** [comment]
> Every day my web server indexes new articles and they get added to this [file](https://github.com/OmnesRes/prepub/blob/master/biorxiv/biorxiv.txt).

Cool, `https://github.com/OmnesRes/prepub/raw/master/biorxiv/biorxiv.txt` is what I need.

Have you considered storing `biorxiv.txt` as JSON? ~~Or at least a TSV?~~ JSON would be better than TSV since you have fields that are lists.

Finally, would it be hard for you to also add the license field to this file? Then this repository wouldn't actually have to download any webpages. In other words, is there any cost for you to add the `bio_license=soup.find('span',{'class':'license-type'}).text` part to your pipeline?

**Comment by OmnesRes** [comment]
I'm going to be honest, I'm not a professional web developer (I just watched some youtube videos).  It is entirely possible a different format might be better, but I've grown to like the format I use, I even have a video about it: https://www.youtube.com/watch?v=OI6Lke65lj8

This format lets me quickly load the data back into Python with [eval(i) for i in f].  Yes, I know pickle can be used for this purpose, but the format I use has the benefit that it is at least somewhat readable.  Every couple days I look at my update files in this format to make sure the articles are getting indexed correctly.  If the format was in JSON or pickled I don't know that I would be able to quickly spot errors.  The format also has the benefit of storing and writing unicode.

There's no reason PrePubMed couldn't start indexing licenses, it could even generate the TSV file for you.

**Comment by dhimmel** [comment]
> There's no reason PrePubMed couldn't start indexing licenses

That would be great. Will hold off scripting processing of `OmnesRes/prepub` data until this is added.

> it could even generate the TSV file for you.

I'm happy to generate the TSV. If you like how I implement it, you can incorporate it into `OmnesRes/prepub`.

**Comment by OmnesRes** [comment]
I think you will be pleased with this [file](https://github.com/OmnesRes/prepub/blob/master/biorxiv/biorxiv_licenses.tsv)

It is similar to the file you generated, but also has authors (ASCII characters only), titles, and affiliations.

It is UTF-8 encoded.

Theoretically PrePubMed will add to the file once a day...or maybe something terrible will happen, we'll have to wait and see.

Although the file gets changed once a day obviously the file in GitHub only gets updated when I perform a commit.  Hopefully the frequency of my update commits will be satisfactory.

Working on projects with other people is fun.

**Comment by dhimmel** [comment]
> Although the file gets changed once a day obviously the file in GitHub only gets updated when I perform a commit. Hopefully the frequency of my update commits will be satisfactory.

What are you using to schedule and run daily PrePubMed updates? You potentially could also have that service automatically commit back to GitHub if successful. See for example [`dhimmel/thinklytics`](https://github.com/dhimmel/thinklytics#scheduled-export) which downloads Thinklab content and commits the results, once a day. Warning, setting this up is probably not [worth the time](https://xkcd.com/1205/ "XKCD Is It Worth the Time?").

**Comment by OmnesRes** [comment]
I have a scheduled PythonAnywhere task that runs daily.  I like to check the files before I make commits.

**Comment by OmnesRes** [comment]
Okay, I pushed a fix.  The file should be correct now without any spurious tab characters.  I also incorporated your regex.  There is a space between CC and the rest of the license.

**Comment by OmnesRes** [comment]
The daily indexing seems to be working fine.  Are you content with waiting until I make commits for updates to the TSV file or do you want a script that you can run to get the most recent preprints possible?

**Comment by dhimmel** [comment]
> The daily indexing seems to be working fine. Are you content with waiting until I make commits for updates

Yeah, I'm happy just getting the latest version pushed to GitHub (see [`1.download.ipynb`](https://github.com/dhimmel/biorxiv-licenses/blob/4162a6d366e5c4791c1ffb1c1077e60bee56b6f7/1.download.ipynb)). Especially since I haven't set up any sort of automated/scheduled updates and don't know if I will.

**Comment by OmnesRes** [comment]
If you want the freshest data I made a commit: https://github.com/OmnesRes/prepub/commit/c19c5b1d4bd2bdcfb2abd70350b0d0c3761a4fa4

I recently cloned my repository to another computer and I was shocked at how large it's gotten.  I realized that the python dictionaries I'm committing are likely causing the repository to balloon in size so I added them to the .gitignore file in a separate commit.  It will make it a little more of a pain when I have to update the SQL database but without this measure I'm going to approach the 1GB limit on GitHub.
